include/appbase/application_base.hpp

```diff
@@ -0,0 +1,327 @@
+            return bogus;
+         }
+
+         template<typename Plugin>
+         Plugin* find_plugin()const {
+            string name = boost::core::demangle(typeid(Plugin).name());
+            return dynamic_cast<Plugin*>(find_plugin(name));
+         }
+
+         template<typename Plugin>
+         Plugin& get_plugin()const {
+            auto ptr = find_plugin<Plugin>();
+            return *ptr;
+         }
+
+         /**
+          * Fetch a reference to the method declared by the passed in type.  This will construct the method
+          * on first access.  This allows loose and deferred binding between plugins
+          *
+          * @tparam MethodDecl - @ref appbase::method_decl
+          * @return reference to the method described by the declaration
+          */
+         template<typename MethodDecl>
+         auto get_method() -> std::enable_if_t<is_method_decl<MethodDecl>::value, typename MethodDecl::method_type&>
+         {
+            using method_type = typename MethodDecl::method_type;
+            auto key = std::type_index(typeid(MethodDecl));
+            auto itr = methods.find(key);
+            if(itr != methods.end()) {
+               return *method_type::get_method(itr->second);
+            } else {
+               methods.emplace(std::make_pair(key, method_type::make_unique()));
+               return  *method_type::get_method(methods.at(key));
+            }
+         }
+
+         /**
+          * Fetch a reference to the channel declared by the passed in type.  This will construct the channel
+          * on first access.  This allows loose and deferred binding between plugins
+          *
+          * @tparam ChannelDecl - @ref appbase::channel_decl
+          * @return reference to the channel described by the declaration
+          */
+         template<typename ChannelDecl>
+         auto get_channel() -> std::enable_if_t<is_channel_decl<ChannelDecl>::value, typename ChannelDecl::channel_type&>
+         {
+            using channel_type = typename ChannelDecl::channel_type;
+            auto key = std::type_index(typeid(ChannelDecl));
+            auto itr = channels.find(key);
+            if(itr != channels.end()) {
+               return *channel_type::get_channel(itr->second);
+            } else {
+               channels.emplace(std::make_pair(key, channel_type::make_unique()));
+               return  *channel_type::get_channel(channels.at(key));
+            }
+         }
+
+         /**
+          * Post func to run on io_service with given priority.
+          *
+          * @param priority can be appbase::priority::* constants or any int, larger ints run first
+          * @param func function to run on io_service
+          * @return result of boost::asio::post
+          */
+         template <typename Executor, typename Func>
+         auto post( Executor &exec, int priority, Func&& func ) {
+            return exec.post(priority, std::forward<Func>(func));
+         }
```

Comment:
Seems like we will want to remove or at least deprecate this. I would vote for just removing it even though it would be a breaking change. Someone can just not update their `appbase` submodule until they are ready to use the new interface. The new interface will be:
`app().executor().post(...)` where `...` I assume will include some read-only vs write enum argument.

Reply:
done, it is removed.

Reply:
Probably we should deprecate it first before removing it.

Reply:
We didn't deprecate anything when we added the `priority` option. I think we should add the new multi-queue option and not try to make it backwards compatible.